##INCOME CLASSIFICATION

*   Author: Satabdi Mishra
*   Date: 01/08/2021



In [1]:
import pandas as pd
import numpy as np
import math
import matplotlib.pyplot as plt
import seaborn as sns
sns.set_style('darkgrid')


In [2]:
dataset = pd.read_csv('data.csv')
df = dataset.copy()

In [3]:
df.head()

age          workclass   fnlwgt  ...  hours-per-week   native-country  income
0   39          State-gov    77516  ...              40    United-States   <=50K
1   50   Self-emp-not-inc    83311  ...              13    United-States   <=50K
2   38            Private   215646  ...              40    United-States   <=50K
3   53            Private   234721  ...              40    United-States   <=50K
4   28            Private   338409  ...              40             Cuba   <=50K

[5 rows x 15 columns]

DATA ANALYSIS

In [4]:
df.describe()

age        fnlwgt  ...   capital-loss   hours-per-week
count  32561.000000  3.256100e+04  ...   32561.000000     32561.000000
mean      38.581647  1.897784e+05  ...      87.303830        40.437456
std       13.640433  1.055500e+05  ...     402.960219        12.347429
min       17.000000  1.228500e+04  ...       0.000000         1.000000
25%       28.000000  1.178270e+05  ...       0.000000        40.000000
50%       37.000000  1.783560e+05  ...       0.000000        40.000000
75%       48.000000  2.370510e+05  ...       0.000000        45.000000
max       90.000000  1.484705e+06  ...    4356.000000        99.000000

[8 rows x 6 columns]

In [5]:
#Null Check

df.isnull().sum()

age                0
 workclass         0
 fnlwgt            0
 education         0
 education-num     0
 marital-status    0
 occupation        0
 relationship      0
 race              0
 sex               0
 capital-gain      0
 capital-loss      0
 hours-per-week    0
 native-country    0
 income            0
dtype: int64

In [6]:
df.dtypes

age                 int64
 workclass         object
 fnlwgt             int64
 education         object
 education-num      int64
 marital-status    object
 occupation        object
 relationship      object
 race              object
 sex               object
 capital-gain       int64
 capital-loss       int64
 hours-per-week     int64
 native-country    object
 income            object
dtype: object

In [7]:
df.columns

Index(['age', ' workclass', ' fnlwgt', ' education', ' education-num',
       ' marital-status', ' occupation', ' relationship', ' race', ' sex',
       ' capital-gain', ' capital-loss', ' hours-per-week', ' native-country',
       ' income'],
      dtype='object')

In [8]:
df[' income'].unique()

array([' <=50K', ' >50K'], dtype=object)

In [9]:
#Encode income variable
cleanup_income = {' income':{' <=50K':0,' >50K':1}}

df = df.replace(cleanup_income)

df.head()

age          workclass   fnlwgt  ...  hours-per-week   native-country  income
0   39          State-gov    77516  ...              40    United-States       0
1   50   Self-emp-not-inc    83311  ...              13    United-States       0
2   38            Private   215646  ...              40    United-States       0
3   53            Private   234721  ...              40    United-States       0
4   28            Private   338409  ...              40             Cuba       0

[5 rows x 15 columns]

In [10]:
df[[' education',' education-num']].value_counts()

 education      education-num
 HS-grad       9                 10501
 Some-college  10                 7291
 Bachelors     13                 5355
 Masters       14                 1723
 Assoc-voc     11                 1382
 11th          7                  1175
 Assoc-acdm    12                 1067
 10th          6                   933
 7th-8th       4                   646
 Prof-school   15                  576
 9th           5                   514
 12th          8                   433
 Doctorate     16                  413
 5th-6th       3                   333
 1st-4th       2                   168
 Preschool     1                    51
dtype: int64

In [11]:
# Drop Irrelevant columns

df.drop(' education',inplace=True, axis=1)
df.head()

age          workclass   fnlwgt  ...   hours-per-week  native-country  income
0   39          State-gov    77516  ...               40   United-States       0
1   50   Self-emp-not-inc    83311  ...               13   United-States       0
2   38            Private   215646  ...               40   United-States       0
3   53            Private   234721  ...               40   United-States       0
4   28            Private   338409  ...               40            Cuba       0

[5 rows x 14 columns]

FEATURE ENGINEERING

In [12]:
X = df.iloc[:, :-1]
y = df.iloc[:, -1]

In [13]:
#Dummify

X = pd.get_dummies(X, drop_first=True)

X.shape

(32561, 85)

In [14]:
#Train test split
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.25, random_state = 1,stratify =y, shuffle = True)

In [15]:
# Feature Scaling
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
X_train = sc.fit_transform(X_train)
X_test = sc.transform(X_test)



Model Imports

In [16]:
from sklearn.metrics import confusion_matrix, accuracy_score
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.svm import SVC
from sklearn.ensemble import RandomForestClassifier


### Logistic Regression

In [17]:
classifier = LogisticRegression(random_state = 0)
classifier.fit(X_train, y_train)

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=100,
                   multi_class='auto', n_jobs=None, penalty='l2',
                   random_state=0, solver='lbfgs', tol=0.0001, verbose=0,
                   warm_start=False)

In [18]:
y_pred = classifier.predict(X_test)

In [19]:
#Scoring the model
cm = confusion_matrix(y_test, y_pred)
print(cm)
accuracy_score(y_test, y_pred)

[[5733  448]
 [ 775 1185]]


0.8497727551897801

### Support Vector Classifier

In [20]:
classifier = SVC(kernel = 'linear', random_state = 0)
classifier.fit(X_train, y_train)

SVC(C=1.0, break_ties=False, cache_size=200, class_weight=None, coef0=0.0,
    decision_function_shape='ovr', degree=3, gamma='scale', kernel='linear',
    max_iter=-1, probability=False, random_state=0, shrinking=True, tol=0.001,
    verbose=False)

In [21]:
y_pred = classifier.predict(X_test)

In [22]:
#Scoring the model
cm = confusion_matrix(y_test, y_pred)
print(cm)
accuracy_score(y_test, y_pred)

[[5741  440]
 [ 784 1176]]


0.8496499201572288

### Decision Tree

In [23]:
classifier = DecisionTreeClassifier(criterion = 'entropy', random_state = 0)
classifier.fit(X_train, y_train)

DecisionTreeClassifier(ccp_alpha=0.0, class_weight=None, criterion='entropy',
                       max_depth=None, max_features=None, max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, presort='deprecated',
                       random_state=0, splitter='best')

In [24]:
y_pred = classifier.predict(X_test)

In [25]:
#Scoring the model
cm = confusion_matrix(y_test, y_pred)
print(cm)
accuracy_score(y_test, y_pred)

[[5449  732]
 [ 741 1219]]


0.8190639970519592

### Random Forest

In [26]:
classifier = RandomForestClassifier(n_estimators = 10, criterion = 'entropy', random_state = 0)
classifier.fit(X_train, y_train)

RandomForestClassifier(bootstrap=True, ccp_alpha=0.0, class_weight=None,
                       criterion='entropy', max_depth=None, max_features='auto',
                       max_leaf_nodes=None, max_samples=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, n_estimators=10,
                       n_jobs=None, oob_score=False, random_state=0, verbose=0,
                       warm_start=False)

In [27]:
y_pred = classifier.predict(X_test)


In [28]:
#Scoring the model
cm = confusion_matrix(y_test, y_pred)
print(cm)
accuracy_score(y_test, y_pred)

[[5774  407]
 [ 810 1150]]


0.8505097653850878

Comparing the scores, Random forest gave the best output with accuracy score of 85.05%